sql/analyze.go

```diff
@@ -1083,12 +1083,37 @@ func simplifyComparisonExpr(n *parser.ComparisonExpr) parser.Expr {
 	// <datum>" unless they could not be simplified further in which case
 	// simplifyExpr cannot handle them. For example, "lower(a) = 'foo'"
 	if isVar(n.Left) && isDatum(n.Right) {
-		// All of the comparison operators have the property that when comparing to
-		// NULL they evaluate to NULL (see evalComparisonOp). NULL is not the same
-		// as false, but in the context of a WHERE clause, NULL is considered
-		// not-true which is the same as false.
 		if n.Right == parser.DNull {
-			return parser.DBool(false)
+			switch n.Operator {
+			case parser.Is, parser.IsNotDistinctFrom:
+				// IS NULL or IS {,NOT} DISTINCT FROM NULL expressions. These are valid
+				// comparisons to NULL which return boolean true or false. We're not
+				// utilizing them in index selection (yet), so return true.
+				return parser.DBool(true)
+			case parser.IsDistinctFrom:
+				switch n.Left.(type) {
+				case *qvalue:
+					// Transform "a IS DISTINCT FROM NULL" into "a IS NOT NULL".
+					return &parser.ComparisonExpr{
+						Operator: parser.IsNot,
+						Left:     n.Left,
+						Right:    n.Right,
+					}
+				}
+			case parser.IsNot:
+				switch n.Left.(type) {
+				case *qvalue:
+					// "a IS NOT NULL" can be used during index selection to restrict the
+					// range scanned to keys > NULL.
+					return n
+				}
+			default:
+				// All of the remaining comparison operators have the property that when
+				// comparing to NULL they evaluate to NULL (see evalComparisonOp). NULL is
+				// not the same as false, but in the context of a WHERE clause, NULL is
```

Comment:
I'm sure the documentation comments are insufficient in this code. On the other hand, if you tried to call this code on non-WHERE clause expressions various tests would break all over the place.